ReplicationOperationTests.testRetryTransientReplicationFailure failing assertions

**Build scan**: [Most recent equals failure](https://gradle-enterprise.elastic.co/s/6ziprwtk7t744/console-log?anchor=4205), [Previous failure](https://gradle-enterprise.elastic.co/s/gyekfufhbpbia), [Previous failure](https://gradle-enterprise.elastic.co/s/6yq4lopxmeiys)

**Repro line**: 
```
./gradlew ':server:test' --tests "org.elasticsearch.action.support.replication.ReplicationOperationTests.testRetryTransientReplicationFailure" -Dtests.seed=52001DFB0F7AE8A5 -Dtests.security.manager=true -Dtests.locale=cs-CZ -Dtests.timezone=PST -Druntime.java=11
```

**Reproduces locally?**: Attempted to reproduce locally. Ran several attempts at 1000 runs at a time, and one 10000 iteration, but could not get either of the assertion failures to trip.

**Applicable branches**: master

**Failure history**: [In the last 6 months](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-6M,mode:quick,to:now))&_a=(columns:!(_source),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:e58bf320-7efd-11e8-bf69-63c8ef516157,key:project,negate:!f,params:(query:elasticsearch,type:phrase),type:phrase,value:elasticsearch),query:(match:(project:(query:elasticsearch,type:phrase)))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:e58bf320-7efd-11e8-bf69-63c8ef516157,key:test,negate:!f,params:(query:testRetryTransientReplicationFailure,type:phrase),type:phrase,value:testRetryTransientReplicationFailure),query:(match:(test:(query:testRetryTransientReplicationFailure,type:phrase))))),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'class:%22org.elasticsearch.action.support.replication.ReplicationOperationTests%22'),sort:!(time,desc))), this test case has contributed to 44% of the failures related to this class, though its failure rate over all tests (only 4 so far) is very low. 
<!--
Link to build stats and possible indication of when this started failing and how often it fails
<https://build-stats.elastic.co/app/kibana>
-->
**Failure excerpt**:
```
org.elasticsearch.action.support.replication.ReplicationOperationTests > testRetryTransientReplicationFailure FAILED |  
-- | --
  | java.lang.AssertionError: |  
  | Expected: <{TZh80u64RU6PVU7fRmp2FQ=8119599783513383472, nZDUN9u7RWmdtYrbAPxipQ=-7826599229229376521, HwbF0EovRYKLrvhaVwRPcA=-202613608617851446, whpHFdGlSIuYzz_6eOh-fw=845263349823329646}> |  
  | but: was <{TZh80u64RU6PVU7fRmp2FQ=8119599783513383472, nZDUN9u7RWmdtYrbAPxipQ=-7826599229229376521, HwbF0EovRYKLrvhaVwRPcA=-202613608617851446, whpHFdGlSIuYzz_6eOh-fw=845263349823329646}> |  
  | at __randomizedtesting.SeedInfo.seed([52001DFB0F7AE8A5:2607CB2DF0934A27]:0) |  
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18) |  
  | at org.junit.Assert.assertThat(Assert.java:956) |  
  | at org.junit.Assert.assertThat(Assert.java:923) |  
  | at org.elasticsearch.action.support.replication.ReplicationOperationTests.testRetryTransientReplicationFailure(ReplicationOperationTests.java:239) |  
  | REPRODUCE WITH: ./gradlew ':server:test' --tests "org.elasticsearch.action.support.replication.ReplicationOperationTests.testRetryTransientReplicationFailure" -Dtests.seed=52001DFB0F7AE8A5 -Dtests.security.manager=true -Dtests.locale=cs-CZ -Dtests.timezone=PST -Druntime.java=11
```

```
org.elasticsearch.action.support.replication.ReplicationOperationTests > testRetryTransientReplicationFailure FAILED |  
  | java.lang.AssertionError: |  
  | Expected: <6058662276597524643L> |  
  | but: was null |  
  | at __randomizedtesting.SeedInfo.seed([FF33340B79E6C851:8B34E2DD860F6AD3]:0) |  
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18) |  
  | at org.junit.Assert.assertThat(Assert.java:956) |  
  | at org.junit.Assert.assertThat(Assert.java:923) |  
  | at org.elasticsearch.action.support.replication.ReplicationOperationTests.testRetryTransientReplicationFailure(ReplicationOperationTests.java:238) |  
  | REPRODUCE WITH: ./gradlew ':server:test' --tests "org.elasticsearch.action.support.replication.ReplicationOperationTests.testRetryTransientReplicationFailure" -Dtests.seed=FF33340B79E6C851 -Dtests.security.manager=true -Dtests.locale=es-PY -Dtests.timezone=America/Port_of_Spain -Druntime.java=11 |   
```

--- a/server/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java
@@ -561,8 +561,8 @@ public class ReplicationOperationTests extends ESTestCase {
         final long maxSeqNoOfUpdatesOrDeletes;
         final Supplier<ReplicationGroup> replicationGroupSupplier;
         final PendingReplicationActions pendingReplicationActions;
-        final Map<String, Long> knownLocalCheckpoints = new HashMap<>();
-        final Map<String, Long> knownGlobalCheckpoints = new HashMap<>();
+        final Map<String, Long> knownLocalCheckpoints = Collections.synchronizedMap(new HashMap<>());
+        final Map<String, Long> knownGlobalCheckpoints = Collections.synchronizedMap(new HashMap<>());
 
         TestPrimary(ShardRouting routing, Supplier<ReplicationGroup> replicationGroupSupplier, ThreadPool threadPool) {
             this.routing = routing;

